have_timegm.1
.TH have_timegm 1 "7 September 2022" "have_timegm.sh" "IOCCC tools"
.SH NAME
have_timegm \- test compiler for \fBtimegm(3)\fP, \fBstrptime(3)\fP and \fBstrftime(3)\fP
.SH SYNOPSIS
\fBhave_timegm.sh\fP [\-h] [\-v level]
.SH DESCRIPTION
\fBhave_timegm.sh\fP will attempt to compile and run a test program \fBhave_timegm.c\fP to determine if the compiler can use the functions \fBtimegm(3)\fP, \fBstrptime(3)\fP and \fBstrftime(3)\fP.
If the system compiler does not work we attempt to compensate by directly declaring the prototype for \fBtimegm(3)\fP to let such systems use the tools in order to submit an IOCCC entry.
We will stop trying to do this in 2024.
.SH OPTIONS
.PP
\fB\-h\fP
Show help and exit.
.PP
\fB\-v\fP
Set verbosity level.
.SH EXIT STATUS
.PP
.br
    0    timestamp updated
.br
    1    failed Phase 0 of verification
.br
    2    failed Phase 1 of verification
.br
    3    failed Phase 2 of verification
.br
    4    failed Phase 3 of verification
.br
    5    failed Phase 4 of verification
.br
    6    rpl failed to modify limit_ioccc.h
.br
    7    limit_ioccc.h not found or not readable and writable
.br
    8    -h and help string printed or -V and version string printed
.br
    9    Command line usage error
.br
    >=10  internal error
.SH NOTES
.PP
While there is no bug in the tool we are aware of one issue.
The fact that the tool has to exist in 2022, a full 11 years after C11, is that issue.
We don't mean to blame or shame but this was originally for CentOS 7 (CentOS 8 met an unduly EOL and < 7 already reached their healthy 10 years).
Nevertheless CentOS was designed for stability so we can perhaps forgive them for this problem.
However in 2024 we will stop trying to compensate CentOS and fortunately (for entering the IOCCC at least) CentOS 7 will reach its EOL then so this should not be a problem.
.PP
.SH EXAMPLES
.PP
.nf
Running the script under CentOS 7 will result in:

.RS
\fB
  $ ./have_timegm.sh 
  -DTIMEGM_PROBLEM\fP
.fi
.RE
.PP
.nf
Running the program under macOS and any other more recent OS:

.RS
\fB
 $ ./have_timegm.sh
 $\fP
.fi
.RE
.SH SEE ALSO
.PP
\fBhave_timegm.c\fP


CHANGES.md
# Major changes to the IOCCC entry toolkit


## Release 0.6 2022-09-02

Released IOCCC entry toolkit v0.6 2022-09-02

Updated CHANGES.md for v0.6 2022-09-02.

Changed `MIN_TIMESTAMP` from 1655958810 to 1662145368.

Updated `mkiocccentry` from "0.40 2022-03-15" to "0.41 2022-09-02".
Updated `.info.json` version from "1.10 2022-06-22" to "1.11 2022-09-02".
Updated `.author.json` version from "1.13 2022-06-22" to "1.14 2022-09-02".

Improved code to use new facilities for output to a buffer from dbg release of
v2.5 2022-07-23.

The `chk_foo()` functions in `chk_validate.c` and the `test_foo()` functions in
`entry_util.c` are code complete, although they remain untested and unused.  The
`chkentry` tool, however, is not yet code complete. Later releases of tested
JSON semantic code will no doubt modify these functions.

Improved a number of the ways that JSON field values are checked.  In a number
of cases, code form `mkiocccentry.c` was moved into `test_foo()` functions so
that they could be used by other tools such as the JSON semantic test code.

Added code to generate JSON semantic tables from JSON reference files for
`.info.json` and `.author.json`.  The `jsemcgen.sh` tool manages this by way of
the `jsemtblgen` code generator and header, patch and trailer files (see
`chk.auth.*` and `chk.info.*` files).

Avoided the appearance of attacking any particular individual.  It was not our
intention to disrespect anyone, even though we disagree with some of the
technical decisions.  Where we have fundamental technical disagreements, we
attempted to express those technical disagreements with humour in hopefully a
more fun way.  As also now apologize for how `bison` and `flex` generated code
may look, instead of simply calling it ugly.  As such, we hopefully improved
some of the humour in our code while trying to be nice and friendly to others.

For example now the adjusted dbg levels in JSON parser are:

```
*     At -J 3, only the top level return type and top level tree are printed.
*     At -J 5, intermediate tree return types and tree are printed too.
*     At -J 7, also print grammar progress.
*     At -J 9, also print sorry_text and sorry_lang grammar values.
```

Removed a number of files and added a number of files under the `test_JSON/`
tree.  When the JSON semantic code is being tested in a future release, we
expect more such `test_JSON/` tree changes.

Improved / add a number of man pages.  Updated `README.md`.

Improved and expanded `txzchk`.

Added more test code.  We attempt to detect feathers in tarballs.  :-)

We will neither confirm nor deny the presence of an "Easter egg".
To to so would be "foolish".  :-)

Improved and fixed `vermod.sh` and `reset_tstamp.sh`.  Tested this
code by changing the `MIN_TIMESTAMP` as noted above.  The `MIN_TIMESTAMP`
needed to up updated anyway due to changes in the `.info.json` and
`.author.json` formats.

Made numerous improvements and bug fixes to the Makefile.

Fixed how `picky` is used in by the `make picky` rule for a few special files.

Added multiple rules to the Makefile including but not limited to `make mkref`,
`make legacy_clobber`, and `make legacy_clean` rules.  Applied multiple bug
fixes to the Makefile.

Improved the Makefile to be less impacting for modern systems while trying to
maintain, for as long as we can, compatibility with some older systems.

Attempted to improve compatibility with reasonably modern systems.  We managed
to keep CentOS 7 somewhat supported for now, although we may be forced to drop
support of such an old system before 2024. But if we do reach 2024 we can only
say that given that CentOS 7 will reach EOL in 2024 and CentOS 8 support was
unduly dropped and since < CentOS 7 already was at EOL this should not be a
problem. :-)

Users of very out of date systems can still enter and submit entries to the
IOCCC.  They just might need to find a more modern system to package and submit
their IOCCC entry, however.

Added stub code for `hostchk.sh`.  A future release will include mode tests for
the given hosts.  Future releases will also include a bug report system that
will also use `hostchk.sh`.

Improved the `no-comment` directive in `.info.json` and `.author.json` files.

Improved how `time_t` values are used and printed.  We no longer assume that
`time_t` is signed nor assume it is unsigned: we only assume it's time. :-)

Improved comments in C code about special `seqcexit` comment directives.

Make numerous bug fixes and fixed a fair number of typos.  Nevertheless we claim
this is *NOT* complete. :-)


## Release 0.5 2022-07-10

The `jsemtblgen` tool is code complete.

Added `is_reserved()` function to `rule_count.c`.

Changed the Official iocccsize version from 28.10 2022-03-15 to 28.11
2022-07-10. Updated `test_JSON` tree and `json_teststr.txt` to reflect the new
`IOCCCSIZE_VERSION` value.

Added and improved a number of man pages.

Added `Makefile` rule to build HTML files from all man pages.

Both the `parse_json()` and `parse_json_file()` top level functions of the JSON
parser return a JSON node parse tree when the JSON is valid.

Trimmed the JSON parser to not include code that does not strictly belong to the
generic JSON parser.  Removed dead C code.

Updated dbg facility to version 2.4 2022-06-26.  Version 2.4 2022-06-26 contains
functions to write (form) a message, debug diagnostic message, warning, or error
message into a string buffer:

```c
extern void snmsg(char *str, size_t size, char const *fmt, ...)
extern void sndbg(char *str, size_t size, int level, char const *fmt, ...);
extern void snwarn(char *str, size_t size, char const *name, char const *fmt, ...);
extern void snwarnp(char *str, size_t size, char const *name, char const *fmt, ...);
extern void snwerr(int error_code, char *str, size_t size, char const *name, char const *fmt, ...);
extern void snwerrp(int error_code, char *str, size_t size, char const *name, char const *fmt, ...);
```


have_timegm.sh
#!/usr/bin/env bash
#
# have_timegm.sh - see if we can compile have_timegm.c
#
# If we can compile have_timegm.c then we will output nothing and exit 0.
#
# If we encounter a compile warning/error while trying to compile have_timegm.c
# or if the output file is not executable or returns an error / does not exit 0
# we will output -DTIMEGM_PROBLEM and exit 0.
#
# All other errors will result in a non-zero exit.

# setup
#
export C_SRC="have_timegm.c"
export PROG="./have_timegm"
export V_FLAG="0"

# parse args
#
while getopts :hv: flag; do
    case "$flag" in
    h) echo "usage: $0 [-h] [-v level]" 1>&2
       exit 2
       ;;
    v) V_FLAG="$OPTARG";
       ;;
    \?) echo "invalid option: -$OPTARG" 1>&2
       exit 3
       ;;
    :) echo "option -$OPTARG requires an argument" 1>&2
       exit 4
       ;;
   *)
       ;;
    esac
done

# verify we have source to compile
#
if [[ ! -e $C_SRC ]]; then
    echo "$0: missing have_timegm.c: $C_SRC" 1>&2
    exit 100
fi
if [[ ! -f $C_SRC ]]; then
    echo "$0: have_timegm.c is not a file: $C_SRC" 1>&2
    exit 101
fi
if [[ ! -r $C_SRC ]]; then
    echo "$0: have_timegm.c is not a readable file: $C_SRC" 1>&2
    exit 102
fi

# prep for compile
#
rm -f "$PROG"

# compile test code
#
cc -std=gnu11 -Wall -Wextra -Werror -pedantic "$C_SRC" -o "$PROG" >/dev/null 2>&1
status="$?"
if [[ $status -ne 0 ]]; then
   echo "-DTIMEGM_PROBLEM"
elif [[ ! -x "$PROG" ]]; then
   echo "-DTIMEGM_PROBLEM"

# execute test code
#
else
    "$PROG" 'Mon Aug 15 21:51:43 2022 UTC' >/dev/null 2>&1
    status="$?"
    if [[ $status -ne 0 ]]; then
       echo "-DTIMEGM_PROBLEM"
    fi
fi

# All Done!!! -- Jessica Noll, Age 2
#
exit 0
